### PR TITLE
Add API endpoint to get token for registering runners for repos

### LINKED ISSUE #23643: Provide a programmable way to register a new action runner
### Feature Description

I wish there was an API endpoint I could query to retrieve a token I could register a new runner with. Currently, the only way to do this is to go to the web UI and copy the token from the page.

### LINKED ISSUE #24750: Automated provisioning of Actions runners
### Feature Description

Currently Actions runner tokens are retrieved from the web UI and are one-time use. This makes it difficult to provision runners in an automated fashion for CI and such.

[Woodpecker](https://woodpecker-ci.org/docs/administration/setup) and [Drone](https://docs.drone.io/runner/docker/installation/linux/) accept a symmetic secret via environment variables for both the application and the runners, making automated provisioning zero-touch. Could such a flow be considered for Gitea please?

P.S. Since it's Gitea issuing and validating the tokens it seemed more appropriate to open this issue here, but I'm happy to move or open a companion issue on act_runner instead :+1: 

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by techknowlogick**
This adds a new route to get a token to register new runners for specific repos

Note: This creates a pattern to create new endpoints for generating runner tokens for orgs and globally (which could be done in other PRs)

Closes #23643
Closes #24750

### CODE REVIEW

File: routers/api/v1/repo/runners.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	"net/http"
+
+	actions_model "code.gitea.io/gitea/models/actions"
+	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/util"
+	"code.gitea.io/gitea/services/convert"
+)
+
+// GenerateRunnerToken generate a new runner token for a repository.
+func GenerateRunnerToken(ctx *context.APIContext) {
```
Comment (wolfogre) [inline routers/api/v1/repo/runners.go:17]: Are there any checks to see if Actions are enabled for the instance/org/repo?

**Comment by yardenshoham** [comment]
Closes #23643?

**Comment by delvh** [comment]
Alternatively, what we could do as well is use **one** route that decides itself if `owner` (and `repo`) are set, which one to create:
1. None provided - site wide, doer must be site admin
2. `owner` provided - org wide, doer must be org admin
3. `owner`+`repo` provided - repo wide, doer must be a repo (or org) admin

**Comment by lunny** [comment]
> Alternatively, what we could do as well is use **one** route that decides itself if `owner` (and `repo`) are set, which one to create:
> 
>     1. None provided - site wide, doer must be site admin
> 
>     2. `owner` provided - org wide, doer must be org admin
> 
>     3. `owner`+`repo` provided - repo wide, doer must be a repo admin

It's not explicit. A type field is better.

**Comment by techknowlogick** [comment]
I could probably re-use the `parseScope` logic from the CLI PR

File: routers/api/v1/api.go [hunk 2]
```diff
@@ -1162,6 +1162,11 @@ func Routes(ctx gocontext.Context) *web.Route {
 				}, reqAnyRepoReader())
 				m.Get("/issue_templates", context.ReferencesGitRepo(), repo.GetIssueTemplates)
 				m.Get("/languages", reqRepoReader(unit.TypeCode), repo.GetLanguages)
+
+				// TODO: list runners
+				// TODO: update runner
+				// TODO: delete runner
+				m.Put("/runners", reqToken(auth_model.AccessTokenScopeRepo), reqRepoWriter(unit.TypeActions), repo.GenerateRunnerToken)
```
Comment (delvh) [inline routers/api/v1/api.go:1191]: Is that only `write` or even `admin`?

Comment (delvh) [inline routers/api/v1/api.go:1191]: Additionally, this route should probably be called `/runners/new(-registration)-token` or something similar.

File: routers/api/v1/repo/runners.go (hunk 1 above)
Comment (delvh) [reply routers/api/v1/repo/runners.go:17]: True, we should probably check that.

File: routers/api/v1/repo/runners.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	// - name: repo
+	//   in: path
+	//   description: name of the repo
+	//   type: string
+	//   required: true
+	// responses:
+	//   "200":
+	//     "$ref": "#/responses/ActionRunnerToken"
+	//   "404":
+	//     "$ref": "#/responses/notFound"
```
Comment (delvh) [inline routers/api/v1/repo/runners.go:38]: What about additionally returning a 404 or 403 in case actions are disabled?

**Comment by techknowlogick** [comment]
Rather than having multiple endpoints, and combining this into one endpoint. What should the endpoint be, as /api/v1/actions doesn't exist as a parent, does it makes sense to make it just for runners?

**Comment by plomosits** [comment]
I think this is a very important feature, especially when using Kubernetes. I had the same problem when I wanted to deploy multiple act_runners. 
Currently I've written a small workaround that uses initContainers to automatically generate new tokens per act_runner and insert them into the postgres database.

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: gitea-runner
data:
  init.sh: |-
    #!/usr/bin/env bash
    export GITEA_RUNNER_REGISTRATION_TOKEN=$(cat /opt/gitea/token)
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: gitea-runner
spec:
  replicas: 3
  selector:
    matchLabels:
      app: gitea-runner
  template:
    metadata:
      labels:
        app: gitea-runner
    spec:
      initContainers:
        - name: init-create-new-token
          image: busybox
          command:
            - sh
            - '-c'
            - tr -dc A-Za-z0-9 </dev/urandom | head -c 40 > /opt/gitea/token
          volumeMounts:
            - name: gitea-token
              mountPath: /opt/gitea
        - name: init-add-new-token
          image: governmentpaas/psql
          command:
            - sh
            - '-c'
            - >-
              GITEA_RUNNER_REGISTRATION_TOKEN=$(/bin/cat /opt/gitea/token);
              psql postgresql://$POSTGRES_USER:$POSTGRES_PASSWORD@$POSTGRES_HOST/$POSTGRES_DB -c
              "insert into public.action_runner_token
                (token, owner_id, repo_id, is_active, created, updated) values
                ('$GITEA_RUNNER_REGISTRATION_TOKEN', 0, 0, 'f', FLOOR(EXTRACT(EPOCH FROM current_timestamp)), FLOOR(EXTRACT(EPOCH FROM current_timestamp)));"
          env:
            - name: POSTGRES_HOST
              value: gitea-postgresql
            - name: POSTGRES_DB
              value: gitea
            - name: POSTGRES_USER
              value: gitea
            - name: POSTGRES_PASSWORD
              valueFrom:
                secretKeyRef:
                  name: gitea-postgresql
                  key: password
          volumeMounts:
            - name: gitea-token
              mountPath: /opt/gitea
      containers:
        - name: runner
          image: vegardit/gitea-act-runner:dind-latest
          env:
            - name: INIT_SH_FILE
              value: /opt/runner/init.sh
            - name: GITEA_INSTANCE_URL
              value: http://gitea-http:3000
          securityContext:
            privileged: true
          volumeMounts:
            - name: gitea-token
              mountPath: /opt/gitea
            - name: init-sh
              mountPath: /opt/runner
      volumes:
        - name: gitea-token
          emptyDir: {}
        - name: init-sh
          configMap:
            name: gitea-runner
            items:
              - key: init.sh
                path: init.sh
```

While this works really well, I'm not entirely happy with the solution. It would be really great if you could register a new act_runner without a token via an API. For security reasons, I think it makes sense to use admin credentials for such a call. This could easily be realized by extending the existing RegisterRequest.

```
message RegisterRequest {
  string name = 1;
  oneof token_or_admin_credentials {
    string token = 2;
    Credentials admin_credentials = 5;
  }
  repeated string agent_labels = 3;
  repeated string custom_labels = 4;
}

message Credentials {
  string username = 1;
  string password = 2;
}
```

It would also be great if there was a way to unregister act_runner as well, especially if you want them to be created and deleted dynamically, to avoid having unnecessary act_runner corpses in the database.

What do think?

**Comment by lunny** [comment]
> I could probably re-use the `parseScope` logic from the CLI PR

I mean we can have the type in the API interface, but in fact, stored the tokens like before

**Comment by sebthom** [comment]
Instead of introducing a new API, wouldn't it be more straight forward to introduce a separate scope for registering runners into the regular gitea tokens, e.g. `admin:register_runner` or `register:runner`? What is so special about runner registration tokens?

**Comment by markopolo123** [comment]
I've just turned up looking for the ability to register and deregister runners programmatically, great to see conversation and work already happening on it!

> Instead of introducing a new API, wouldn't it be more straight forward to introduce a separate scope into the regular gitea tokens for registering runners, e.g. `admin:register_runner` or `register:runner`? What is so special about runner registration tokens?
> 
> Now you have to send a HTTP request to gitea (with a gitea token for auth) just to get a runner registration token. This feels like an unnecessary indirection.

my 2c on this (please bare in mind I'm not too familiar yet with the gitea codebase):

I'd like for runner tokens to be short lived, and only capable of use for registering a runner at either repo, team, org or global level. The reason for this is that that token is exposed directly to the runner.

**Comment by christhomas** [comment]
> apiVersion: v1
> kind: ConfigMap
> metadata:
>   name: gitea-runner
> data:
>   init.sh: |-
>     #!/usr/bin/env bash
>     export GITEA_RUNNER_REGISTRATION_TOKEN=$(cat /opt/gitea/token)
> ---
> apiVersion: apps/v1
> kind: Deployment
> metadata:
>   name: gitea-runner
> spec:
>   replicas: 3
>   selector:
>     matchLabels:
>       app: gitea-runner
>   template:
>     metadata:
>       labels:
>         app: gitea-runner
>     spec:
>       initContainers:
>         - name: init-create-new-token
>           image: busybox
>           command:
>             - sh
>             - '-c'
>             - tr -dc A-Za-z0-9 </dev/urandom | head -c 40 > /opt/gitea/token
>           volumeMounts:
>             - name: gitea-token
>               mountPath: /opt/gitea
>         - name: init-add-new-token
>           image: governmentpaas/psql
>           command:
>             - sh
>             - '-c'
>             - >-
>               GITEA_RUNNER_REGISTRATION_TOKEN=$(/bin/cat /opt/gitea/token);
>               psql postgresql://$POSTGRES_USER:$POSTGRES_PASSWORD@$POSTGRES_HOST/$POSTGRES_DB -c
>               "insert into public.action_runner_token
>                 (token, owner_id, repo_id, is_active, created, updated) values
>                 ('$GITEA_RUNNER_REGISTRATION_TOKEN', 0, 0, 'f', FLOOR(EXTRACT(EPOCH FROM current_timestamp)), FLOOR(EXTRACT(EPOCH FROM current_timestamp)));"
>           env:
>             - name: POSTGRES_HOST
>               value: gitea-postgresql
>             - name: POSTGRES_DB
>               value: gitea
>             - name: POSTGRES_USER
>               value: gitea
>             - name: POSTGRES_PASSWORD
>               valueFrom:
>                 secretKeyRef:
>                   name: gitea-postgresql
>                   key: password
>           volumeMounts:
>             - name: gitea-token
>               mountPath: /opt/gitea
>       containers:
>         - name: runner
>           image: vegardit/gitea-act-runner:dind-latest
>           env:
>             - name: INIT_SH_FILE
>               value: /opt/runner/init.sh
>             - name: GITEA_INSTANCE_URL
>               value: http://gitea-http:3000
>           securityContext:
>             privileged: true
>           volumeMounts:
>             - name: gitea-token
>               mountPath: /opt/gitea
>             - name: init-sh
>               mountPath: /opt/runner
>       volumes:
>         - name: gitea-token
>           emptyDir: {}
>         - name: init-sh
>           configMap:
>             name: gitea-runner
>             items:
>               - key: init.sh
>                 path: init.sh

thank you! this is exactly what I needed to fix my problem, even though it would be better if you didn't need to do this

amazing work!

**Pull request closed without merging**